src/leader.c

```diff
@@ -145,41 +145,36 @@ void leader__close(struct leader *l)
 	QUEUE__REMOVE(&l->queue);
 }
 
+/* A checkpoint command that fails to commit is not a huge issue.
+ * The WAL will not be checkpointed this time around on these nodes,
+ * a new checkpoint command will be issued once the WAL on the leader reaches
+ * threshold size again. It's improbable that the WAL in this way could grow
+ * without bound, it would mean that apply frames commands commit without
+ * issues, while the checkpoint command would somehow always fail to commit. */
 static void leaderCheckpointApplyCb(struct raft_apply *req,
 				    int status,
 				    void *result)
 {
-	struct leader *l = req->data;
+	(void)req;
 	(void)result;
-	/* In case of failure, release the chekcpoint lock. */
 	if (status != 0) {
                 tracef("checkpoint apply failed %d", status);
-		struct sqlite3_file *file;
-		sqlite3_file_control(l->conn, "main", SQLITE_FCNTL_FILE_POINTER,
-				     &file);
-		file->pMethods->xShmLock(
-		    file, 1 /* checkpoint lock */, 1,
-		    SQLITE_SHM_UNLOCK | SQLITE_SHM_EXCLUSIVE);
 	}
-	l->inflight = NULL;
-	l->db->tx_id = 0;
-	l->exec->done = true;
-	maybeExecDone(l->exec);
 }
 
-/* Attempt to perform a checkpoint if possible. */
-static bool leaderMaybeCheckpoint(struct leader *l)
+/* Attempt to perform a checkpoint on nodes running a version of dqlite that
+ * doens't perform autonomous checkpoints. For recent nodes, the checkpoint
+ * command will just be a no-op.
+ * This function will run after the WAL might have been checkpointed during a call
+ * to `apply_frames`.
+ * */
+static void leaderMaybeCheckpointLegacy(struct leader *l)
```

Comment:
Is this really necessary? It sounds like it will generate a checkpoint command even on clusters having all new nodes, where that's not needed. I believe that the situation where the leader is newer and there are older followers should be generally short lived (since those older followers should get updated soon), and maybe it makes more sense to optimize for the common case where all nodes are new.

Reply:
I know it's not ideal but on a write heave database the WAL can quickly grow out of control I think. I could leave this in for a couple of months and remove it after say 4-6 months.